adapters/Maio/MaioAdapter/GADMMaioInterstitialAdapter.m

```diff
@@ -148,81 +123,47 @@ - (BOOL)isBannerAnimationOK:(GADMBannerAnimationType)animType {
 /// and adapterDidDismissInterstitial: when the interstitial is being dismissed.
 - (void)presentInterstitialFromRootViewController:(UIViewController *)rootViewController {
   [self.interstitialAdConnector adapterWillPresentInterstitial:self];
-  GADMAdapterMaioAdsManager *adManager =
-      [GADMAdapterMaioAdsManager getMaioAdsManagerByMediaId:self.mediaId];
-  [adManager showAdForZoneId:self.zoneId rootViewController:rootViewController];
-}
-
-#pragma mark - MaioDelegate
-
-/**
- *  全てのゾーンの広告表示準備が完了したら呼ばれます。
- */
-- (void)maioDidInitialize {
-  // noop
+  [self.interstitial showWithViewContext:rootViewController callback:self];
 }
 
-/**
- *  広告の配信可能状態が変更されたら呼ばれます。
- *
- *  @param zoneId   広告の配信可能状態が変更されたゾーンの識別子
- *  @param newValue 変更後のゾーンの状態。YES なら配信可能
- */
-- (void)maioDidChangeCanShow:(NSString *)zoneId newValue:(BOOL)newValue {
-  if (!newValue) return;
+#pragma mark - MaioInterstitialLoadCallback, MaioInterstitialShowCallback
 
+- (void)didLoad:(MaioInterstitial *)ad {
   [self.interstitialAdConnector adapterDidReceiveInterstitial:self];
-}
 
-/**
- *  広告が再生される直前に呼ばれます。
- *  最初の再生開始の直前にのみ呼ばれ、リプレイ再生の直前には呼ばれません。
- *
- *  @param zoneId  広告が表示されるゾーンの識別子
- */
-- (void)maioWillStartAd:(NSString *)zoneId {
-  // NOOP
 }
 
-/**
- *  広告の再生が終了したら呼ばれます。
- *  最初の再生終了時にのみ呼ばれ、リプレイ再生の終了時には呼ばれません。
- *
- *  @param zoneId  広告を表示したゾーンの識別子
- *  @param playtime 動画の再生時間（秒）
- *  @param skipped  動画がスキップされていたら YES、それ以外なら NO
- *  @param rewardParam
- * ゾーンがリワード型に設定されている場合、予め管理画面にて設定してある任意の文字列パラメータが渡されます。それ以外の場合は
- * nil
- */
-- (void)maioDidFinishAd:(NSString *)zoneId
-               playtime:(NSInteger)playtime
-                skipped:(BOOL)skipped
-            rewardParam:(NSString *)rewardParam {
-  // NOOP
-}
-
-/**
- *  広告がクリックされ、ストアや外部リンクへ遷移した時に呼ばれます。
- *
- *  @param zoneId  広告がクリックされたゾーンの識別子
- */
-- (void)maioDidClickAd:(NSString *)zoneId {
+- (void)didFail:(MaioInterstitial *)ad errorCode:(NSInteger)errorCode {
   id<GADMAdNetworkConnector> strongConnector = self.interstitialAdConnector;
   if (!strongConnector) {
     return;
   }
+  NSString *description = @"maio SDK returned error";
+  NSDictionary *userInfo =
+      @{NSLocalizedDescriptionKey : description, NSLocalizedFailureReasonErrorKey : description};
+  NSError *error = [NSError errorWithDomain:GADMMaioSDKErrorDomain
+                                       code:errorCode
+                                   userInfo:userInfo];
+
+  NSLog(@"MaioInterstitial did fail. error: %@", error);
+
+  if (10000 <= errorCode && errorCode < 20000) {
+    // Fail to load.
+    [strongConnector adapter:self didFailAd:error];
+  } else if (20000 <= errorCode && errorCode < 30000) {
+    // Fail to Show
+    [strongConnector adapter:self didFailAd:error];
+  } else {
+    // Unknown error code
+    [strongConnector adapter:self didFailAd:error];
+  }
+}
 
-  [strongConnector adapterDidGetAdClick:self];
-  [strongConnector adapterWillLeaveApplication:self];
+- (void)didOpen:(MaioInterstitial *)ad {
+  // NOOP
```

Comment:
Shouldn't we call `adapterWillPresentInterstitial` here?

Is this callback not being called or it's just GMA SDK doesn't have the equivalent? If it's latter, it should be `  // Google Mobile Ads SDK doesn't have a matching event.`

Reply:
@im-ttmskk Is there an update on this?

Reply:
It is my understanding that `adapterWillPresentInterstitial` notifications are used by developers to avoid conflicting with ad behavior (e.g. to turn off app sound).
Ads are already displayed on `didOpen: . So I don't think it's appropriate.
Therefore, I will use the comments you provided.